Align the directionality of bdi and input elements with HTML5 spec

rniwa opened the pull request:
#### 92c7ebfc2fbb17b42e983e554890966dce8469a2
<pre>
Align the directionality of bdi and input elements with HTML5 spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=243130">https://bugs.webkit.org/show_bug.cgi?id=243130</a>

Reviewed by Cameron McCormack.

Align WebKit&apos;s behavior with HTML5 specification:
<a href="https://html.spec.whatwg.org/multipage/dom.html#the-directionality">https://html.spec.whatwg.org/multipage/dom.html#the-directionality</a>

Specifically, bdi element with invalid dir content attribute should behave as if dir=auto is set,
and only input element whose type attribute is in text, search, telephone, URL, or email state
should examine its value to determine the directionality.

* LayoutTests/fast/css/bdi-element-invalid-dir-expected.html: Added.
* LayoutTests/fast/css/bdi-element-invalid-dir.html: Added. Firefox passes but Chrome fails.
* LayoutTests/fast/css/dir-pseudo-on-bdi-element-expected.txt: Added.
* LayoutTests/fast/css/dir-pseudo-on-bdi-element.html: Added. Firefox pass all test cases.
* LayoutTests/fast/css/dir-pseudo-on-input-element-expected.txt: Added.
* LayoutTests/fast/css/dir-pseudo-on-input-element.html: Added. Firefox passes all test cases
except the one with input element with type=tel inside a RTL block.
<a href="https://bugzilla.mozilla.org/show_bug.cgi?id=1780917">https://bugzilla.mozilla.org/show_bug.cgi?id=1780917</a> tracks the bug in Gecko.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/dir-expected.txt:
* LayoutTests/platform/mac-wk1/fast/css/dir-pseudo-on-input-element-expected.txt: Added.
* LayoutTests/platform/win/TestExpectations:

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::isValidDirValue): Added.
(WebCore::HTMLElement::hasDirectionAuto const): Fixed the bug that bdi with an invalid dir
doesn&apos;t behave as if dir is omitted.
(WebCore::HTMLElement::computeDirectionality const): Added a special case for telephone.
(WebCore::HTMLElement::directionality const): Only use the input element&apos;s value to determine
the directionality of the element only if its type is text, search, telephone, URL, or email state.

* Source/WebCore/html/HTMLInputElement.h:
(WebCore::HTMLInputElement): Made isTextType() public to be called in HTMLElement::directionality.

Canonical link: <a href="https://commits.webkit.org/252779@main">https://commits.webkit.org/252779@main</a>
</pre>

nt1m reviewed (commented):
What prevents the new tests from being WPT? You're using the harness already :)

heycam commented on LayoutTests/fast/css/bdi-element-invalid-dir.html:
```diff
@@ -0,0 +1,16 @@
+</style>
+</head>
+<body>
+<div><bdi dir="foo"><span id="red"></span><span>&#x05EA;</span></bdi></div>
```
It's confusing that the `span` is `id="red"`, since it isn't red itself. (It's covering the left half of the red `div`.) Maybe this could be `id="left"`?

rniwa replied:
Yeah, the meaning of this span changed when I fixed iOS simulator failure. Will rename it to left.

heycam reviewed (approved):
r=me

rniwa committed:
Align the directionality of bdi and input elements with HTML5 spec
https://bugs.webkit.org/show_bug.cgi?id=243130

Reviewed by Cameron McCormack.

Align WebKit's behavior with HTML5 specification:
https://html.spec.whatwg.org/multipage/dom.html#the-directionality

Specifically, bdi element with invalid dir content attribute should behave as if dir=auto is set,
and only input element whose type attribute is in text, search, telephone, URL, or email state
should examine its value to determine the directionality.

* LayoutTests/fast/css/bdi-element-invalid-dir-expected.html: Added.
* LayoutTests/fast/css/bdi-element-invalid-dir.html: Added. Firefox passes but Chrome fails.
* LayoutTests/fast/css/dir-pseudo-on-bdi-element-expected.txt: Added.
* LayoutTests/fast/css/dir-pseudo-on-bdi-element.html: Added. Firefox pass all test cases.
* LayoutTests/fast/css/dir-pseudo-on-input-element-expected.txt: Added.
* LayoutTests/fast/css/dir-pseudo-on-input-element.html: Added. Firefox passes all test cases
except the one with input element with type=tel inside a RTL block.
https://bugzilla.mozilla.org/show_bug.cgi?id=1780917 tracks the bug in Gecko.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/selectors/pseudo-classes/dir-expected.txt:
* LayoutTests/platform/mac-wk1/fast/css/dir-pseudo-on-input-element-expected.txt: Added.
* LayoutTests/platform/win/TestExpectations:

* Source/WebCore/html/HTMLElement.cpp:
(WebCore::isValidDirValue): Added.
(WebCore::HTMLElement::hasDirectionAuto const): Fixed the bug that bdi with an invalid dir
doesn't behave as if dir is omitted.
(WebCore::HTMLElement::computeDirectionality const): Added a special case for telephone.
(WebCore::HTMLElement::directionality const): Only use the input element's value to determine
the directionality of the element only if its type is text, search, telephone, URL, or email state.

* Source/WebCore/html/HTMLInputElement.h:
(WebCore::HTMLInputElement): Made isTextType() public to be called in HTMLElement::directionality.

Canonical link: https://commits.webkit.org/252779@main
